src/main/java/com/limechain/trie/structure/nibble/Nibbles.java

```diff
@@ -92,191 +84,148 @@ public boolean startsWith(Nibbles prefix) {
+     * @return the new Nibbles after insertion
+     * @implNote will create a new modified copy of the nibbles, does not mutate the instance invoked on
+     */
+    public Nibbles add(Nibble nibble) {
+        List<Nibble> newNibbles = new ArrayList<>(this.nibbles);
+        newNibbles.add(nibble);
+        return new Nibbles(newNibbles);
     }
 
-    @Override
-    public void clear() {
-        nibbles.clear();
+    /**
+     * Adds a new nibble to the nibbles at the given position.
+     * @param nibble the new nibble to add
+     * @return the new Nibbles after insertion
+     * @implNote will create a new modified copy of the nibbles, does not mutate the instance invoked on
+     */
+    public Nibbles add(int index, Nibble nibble) {
+        List<Nibble> newNibbles = new ArrayList<>(this.nibbles);
+        newNibbles.add(index, nibble);
+        return new Nibbles(newNibbles);
     }
 
-    @Override
+    /**
+     * Returns the element at the specified position in this list.
+     *
+     * @param index index of the element to return
+     * @return the element at the specified position in this list
+     * @throws IndexOutOfBoundsException if the index is out of range
+     *         ({@code index < 0 || index >= size()})
+     */
     public Nibble get(int index) {
         return nibbles.get(index);
     }
 
-    @Override
-    public Nibble set(int index, Nibble element) {
-        return nibbles.get(index);
+    // NOTE:
+    //  Since Nibbles is immutable, we don't actually need to explicitly copy
+    //  This method could go away, passing by reference could suffice, although going back to mutability
+    //  if later decided would become rather difficult.
+    public Nibbles copy() {
+        return Nibbles.of(this.nibbles);
     }
 
-    @Override
-    public void add(int index, Nibble element) {
-        nibbles.add(index, element);
+    /**
+     * Drops a number of nibbles from the beginning
+     * @param n the number of nibbles to skip
+     * @return a new Nibbles without the first n leading nibbles
+     * @throws IndexOutOfBoundsException if
+     *         ({@code n < 0 || n > size})
+     */
+    public Nibbles drop(int n) {
+        // NOTE: We might want to not throw, but return an empty Nibbles instead if (n < 0 || n > size).
+        return new Nibbles(this.nibbles.subList(n, this.nibbles.size()));
     }
 
-    @Override
-    public Nibble remove(int index) {
-        return nibbles.remove(index);
+    /**
+     * Takes only the first n nibbles from the beginning.
+     * @param n the number of nibbles to take
+     * @return a new Nibbles limited to only the first n leading nibbles
+     * @throws IndexOutOfBoundsException if
+     *         ({@code n < 0 || n > size})
+     */
+    public Nibbles take(int n) {
+        // NOTE: We might want to not throw, but return the same Nibbles instead if (n > size).
+        return new Nibbles(this.nibbles.subList(0, n));
     }
```

Comment:
I couldn't decide on the behaviour I prefer for these two (although it doesn't matter as much, but still).
Since these methods come from rather abundant outside usage of `.stream().skip(n).collect(NibblesCollector.toNibbles())`, it makes sense for them to behave like `Stream`'s skip / take (hence not throwing anything). 
On the other hand, turning those into methods of Nibbles allows us to make use of the underlying list and only copy a view of the sublist we're interested in and simply wrap it into a new nibbles (without unnecessary copying). But `subList` throws exceptions, so it made natural sense to do this as well.

Ultimately, we could add a simple check for the boundaries and act stream-likey while still taking advantage of `subList`.

Reply:
I'm fine with them throwing exceptions.